test/helpers/customError.js

```diff
@@ -1,19 +1,37 @@
 const { config } = require('hardhat');
+const { expect } = require('chai');
 
 const optimizationsEnabled = config.solidity.compilers.some(c => c.settings.optimizer.enabled);
 
 /** Revert handler that supports custom errors. */
-async function expectRevertCustomError(promise, reason) {
+async function expectRevertCustomError(promise, expectedErrorName, args) {
   try {
     await promise;
     expect.fail("Expected promise to throw but it didn't");
   } catch (revert) {
-    if (reason) {
+    if (!Array.isArray(args)) {
+      expect.fail('Expected 3rd array parameter for error arguments');
+    }
+    if (expectedErrorName) {
       if (optimizationsEnabled) {
         // Optimizations currently mess with Hardhat's decoding of custom errors
-        expect(revert.message).to.include.oneOf([reason, 'unrecognized return data or custom error']);
+        expect(revert.message).to.include.oneOf([expectedErrorName, 'unrecognized return data or custom error']);
       } else {
-        expect(revert.message).to.include(reason);
+        const [, error] = revert.message.match(/'(.*)'/);
+        if (!/([A-Z])\w+\(.*\)/g.test(error)) {
+          expect.fail(`Couldn't parse "${error}" as a custom error`);
+        }
+        const [, errorName] = error.match(/(\w+)\(.*\)/);
+        const argMatches = [...error.replace(errorName, '').matchAll(/(0x[0-9A-Fa-f]+|-?\d+|\w+)/g)];
```

Comment:
Note that `0x[0-9A-Fa-f]+` implies `\w+` so that first branch is not necessary.

Reply:
I tried removing `0x[0-9A-Fa-f]+` but it didn't work because `0` itself it's a number. I think I'm not following what you mean, can you clarify?

Reply:
I think Frangio's point is that `\w` is a shortcut for `[0-9a-zA-Z_]` and it should capture any hex string

Reply:
Yeah, removing `0x[0-9A-Fa-f]+` should've worked the same. :thinking:

Reply:
So it was failing because of the order.

- `/(-?\d+|\w+)/` captures `0` and `x...dead`
- `/(\w+|-?\d+)/` does the job.

I updated accordingly.